pkg/operator/metrics/metrics_test.go

```diff
@@ -100,29 +196,138 @@ func TestCredentialsRequests(t *testing.T) {
 
 				// GCP credreq with condition set
 				testCredReqWithConditions(testGCPCredRequest("gignored"), []credreqv1.CredentialsRequestCondition{ignoredCond}),
+
+				testConfigMap(true),
 			},
-			validate: func(accumulator *credRequestAccumulator, t *testing.T) {
-				// total cred requests
-				assert.Equal(t, 3, accumulator.crTotals["aws"])
-				assert.Equal(t, 1, accumulator.crTotals["gcp"])
+			validate: func(t *testing.T) {
+				metricAssert(t, metricCredentialsRequestTotal, "aws", 3)
+				metricAssert(t, metricCredentialsRequestTotal, "gcp", 1)
 
-				// failure conditions should all be zero as CCO is disabled
 				for _, cond := range constants.FailureConditionTypes {
-					assert.Equal(t, 0, accumulator.crConditions[cond])
+					metricAssert(t, metricCredentialsRequestConditions, string(cond), 0)
 				}
 			},
 		},
 	}
 
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			accumulator := newAccumulator(logger)
 
-			for _, cr := range test.credReqs {
-				accumulator.processCR(&cr, test.ccoDisabled)
+			fakeClient := fake.NewFakeClient(test.existingObjects...)
+
+			mc := &Calculator{
+				Client: fakeClient,
+				log:    logger,
+			}
+
+			mc.metricsLoop()
+
+			test.validate(t)
+		})
+	}
+
+}
+
+func metricAssert(t *testing.T, metric *prometheus.GaugeVec, label string, expectedValue int) {
+	val, err := metric.GetMetricWithLabelValues(label)
+	assert.NoError(t, err, "unexpected error getting stored metric value")
+	assert.Equal(t, float64(expectedValue), promtest.ToFloat64(val))
```

Comment:
Best practice for metrics is that you should treat them like debug log statements, which would seldom be unit tested unless they were integral to the application. I think the mentality is that sprinkling instrumentation around should be quick and easy, which it won't if you need to cover it all with unit tests.

I think there could also be issues with the way they're exposed as something like a global singleton where state could leak from test to test. Not sure if the promtest lib you vendored helps there but regardless.

In our case we have some logic in play, but we can unit test that at the accumulator or calculator level instead of going right to prometheus.

Reply:
Yes, I do think there is the potential issue with state from test to test, but since the metric collection is written in a way to clear metrics before calculation, we are protected from this (as the code is currently written). I thought it was nice to get the extra code coverage, but if you think it's not worth it, I can switch back to having the tests wrap only the accumulator.